usaspending_api/search/v2/views/search.py

```diff
@@ -24,64 +25,87 @@ class SpendingOverTimeVisualizationViewSet(APIView):
 
         # build response
-        response = {"group": group, "results": []}
+        response = {'group': group, 'results': []}
 
-        # filter queryset by time
         group_results = OrderedDict()  # list of time_period objects ie {"fy": "2017", "quarter": "3"} : 1000
-        queryset = queryset.order_by("action_date").values("action_date", "federal_action_obligation")
-
-        for trans in queryset:
-            key = {}
-            if group == "fy" or group == "fiscal_year":
-                fy = generate_fiscal_year(trans["action_date"])
-                key = {"fiscal_year": str(fy)}
-            elif group == "m" or group == "month":
-                fy = generate_fiscal_year(trans["action_date"])
-                m = generate_fiscal_month(trans["action_date"])
-                key = {"fiscal_year": str(fy), "month": str(m)}
-            else:  # quarter
-                fy = generate_fiscal_year(trans["action_date"])
-                q = generate_fiscal_period(trans["action_date"])
-                key = {"fiscal_year": str(fy), "quarter": str(q)}
-
-            key = str(key)
-            if group_results.get(key) is None:
-                group_results[key] = trans["federal_action_obligation"]
-            else:
-                if trans["federal_action_obligation"]:
-                    group_results[key] = group_results.get(key) + trans["federal_action_obligation"]
+
+        if group == 'fy' or group == 'fiscal_year':
+
+            fy_set = queryset.values('fiscal_year')\
+                .annotate(federal_action_obligation=Sum('federal_action_obligation'))
+
+            for trans in fy_set:
+                key = {'fiscal_year': str(trans['fiscal_year'])}
+                key = str(key)
+                group_results[key] = trans['federal_action_obligation']
+
+        elif group == 'm' or group == 'month':
+
+            month_set = queryset.annotate(month=ExtractMonth('action_date')) \
+                .values('fiscal_year', 'month') \
+                .annotate(federal_action_obligation=Sum('federal_action_obligation'))
+
+            for trans in month_set:
+                # Make month the 'fiscal month'
+                month = int(trans['month'])
+                fiscal_month = month + 3 if (month >= 1 and month <= 9) else month - 9
+                key = {'fiscal_year': str(trans['fiscal_year']), 'month': str(fiscal_month)}
+                key = str(key)
+                group_results[key] = trans['federal_action_obligation']
+
+        else: # quarterly, take months and add them up
+
+            month_set = queryset.annotate(month=ExtractMonth('action_date')) \
+                .values('fiscal_year', 'month') \
+                .annotate(federal_action_obligation=Sum('federal_action_obligation'))
+
+            for trans in month_set:
+                # Convert month to quarter
+                month = int(trans['month'])
+                if   1 <= month <= 3: quarter = 2  # Jan-Mar
```

Comment:
look at the library `fiscalyear`, that may be able to do most of what you need here: https://pypi.python.org/pypi/fiscalyear/0.1.0